software/bsg_manycore_lib/bsg_tile_group_barrier.hpp

```diff
@@ -0,0 +1,266 @@
+//====================================================================
+// bsg_tile_group_barrier.hpp
+// 02/14/2019, shawnless.xie@gmail.com
+// 02/19/2020, behsani@cs.washington.edu
+//====================================================================
+// The barrier implementation for tile group in manycore
+// Usage:
+//      1. #include "bsg_tile_group_barrier.hpp"
+//      1. bsg_barrier<Y dimension, X dimension> my_barrier;
+//      3. my_barrier.sync(); 
+//
+//Memory Overhead
+//       (2 + X_DIM + 4) + (2 + Y_DIM +4) 
+//      =(12 + X_DIM + Y_DIM) BYTES
+//
+//Worst Case Performance:
+//      1. row sync     :   X_DIM     cycles //all tiles writes to center tile of the row
+//      2. row polling  : 3*X_DIM     cycles // lbu <xx>; beqz <xxx;
+//      3. col sync     :   Y_DIM     cycles //all tiles writes to the center tiles of the col
+//      4. col polling  : 3*Y_DIM     cycles // lbu <xx>; beqz <xx>;
+//      5. col alert        Y_DIM     cycles // store
+//      6. row alert        X_DIM     cycles // store
+//      -----------------------------------------------
+//                        5*( X_DIM + Y_DIM)
+//      For 3x3 group,  cycles = 181, heavy looping/checking overhead for small groups.
+
+#ifndef  BSG_TILE_GROUP_BARRIER_HPP_
+#define  BSG_TILE_GROUP_BARRIER_HPP_
+
+// We need the global bsg_x,bsg_y value.
+#include "bsg_set_tile_x_y.h"
+#include "bsg_manycore.h"
+#include "bsg_mutex.h"
+#include "bsg_manycore.hpp"
+
+
+
+
+
+// Helper funcitons.
+// check if the char array are all non-zeros
+inline void poll_range( int range, volatile const unsigned int *p);
+
+
+template <int BARRIER_X_DIM>
+class bsg_row_barrier {
+public:
+    static constexpr unsigned char    _x_cord_start = 0;
+    static constexpr unsigned char    _x_cord_end = BARRIER_X_DIM - 1;
+    volatile unsigned int     _local_alert = 0;
+    volatile unsigned int     _done_list[ BARRIER_X_DIM ] = {0};
+
+
+    bsg_row_barrier (){};
+
+
+    // Reinitializes the done_list to zero
+    void reset() {
+        for (int i = 0; i < BARRIER_X_DIM; i ++) {
+            unsigned int* done_list_ptr = const_cast<unsigned int*> (&_done_list[i]);
+            *done_list_ptr = 0;
+        }
+        return;
+    };
+
+
+    // send the sync signal to the center tile of the row
+    // executed by all tiles in the group.
+    void sync (unsigned char center_x_cord) {
+        //write to the corresponding done
+        volatile unsigned int *done_list_ptr = const_cast<unsigned int*> (reinterpret_cast<volatile unsigned int*> (bsg_remote_pointer(center_x_cord, bsg_y, &_done_list[bsg_x - _x_cord_start])));
```

Comment:
Isn't _x_cord_start always zero? If it's always zero, is there really a need to have this constant? (I'm assuming this is safe since we are using tile-group EVA).
and we can replace whereever _x_cord_end is used with BARRIER_X_DIM?

Reply:
Yes that makes sense, I was modeling after the legacy code but you're right it's overkill. How does it look now?